src/classes/OPP_OpportunityContactRoles_TDTM.cls

```diff
@@ -289,11 +292,19 @@ public without sharing class OPP_OpportunityContactRoles_TDTM extends TDTM_Runna
             insert listPrimaryOCRForInsert;
         }
 
+        // Create Relationship Soft Credits accordingly
         if (!primaryContactsToOpportunities.isEmpty()) {
             List<sObject> relationshipOCRs
                 = OPP_AutomatedSoftCreditsService.createRelationshipOCRs(primaryContactsToOpportunities);
             dmlWrapper.objectsToInsert.addAll(relationshipOCRs);
         }
+
+        // Create Affiliation Soft Credits accordingly
+        if (!accountsToOpportunities.isEmpty()) {
```

Comment:
During testing, did you come across a scenario that created duplicate OCRs? The Relationship OCRs are created for Opportunities that have an Individual Account; whereas, the Affiliation OCRs are created for Opportunities that have an Organization Account. Given that, the OCRs created would not be duplicated because of the differing Opportunity records. Please share how you were able to create duplicates. I'd like to add it to our tests.

There is the potential for OPP_AutomatedSoftCreditsService.retrieveOpportunityContactRoles() to be called twice. However, unless there is a bulk process with both Individual and Organization Account Opportunities, then, generally, only one call will be made to that method. Mapping Automated Soft Credits would provide insights about its expected usage.  

I agree there are alternative ways to implement this code. In fact, it would have been great to refactor the OPP_OpportunityContactRoles_TDTM class in general, but I know the preference has been to make minimal changes to the code. If you think the OPP_OpportunityContactRoles_TDTM class now needs to be refactored, let's discuss with Mary and Alex.

Reply:
You may be right that we should only do Rels for individual opps, and Affls for organizational opps.  I couldn't remember if that was discussed and resolved or not.  When we discuss with @mpusto make sure to bring up the scenario of a contact who uses their business or foundation account for donations, but may have relationships to a board member who should get soft credited.

I can't say what the frequency of both org opps and individual opps being inserted at the same time.  It may be rare, but in general I just always try to design code optimized for bulk operations assuming a mix.

You are right that we do try to keep minimal changes to existing code to avoid breaking existing functionality.  that said, I'd still want us to consider having the old code use a new shared cache of OCR's to avoid multiple soql calls.  As for doing a complete redesign, I'd be less excited about going that route, unless we found that there are additional issues uncovered in this trigger handler that would be better resolved with more of a redesign.

Reply:
Sounds good. Let's follow up with @mpusto.